include/revng/Pipeline/Runner.h

```diff
@@ -0,0 +1,234 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <string>
+#include <utility>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/Pipeline/ContainerFactorySet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InvalidationEvent.h"
+#include "revng/Pipeline/KindsRegistry.h"
+#include "revng/Pipeline/Structure.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Support/Debug.h"
+
+namespace Pipeline {
+
+/// A runner is a wrapper around a pipeline structure and the context needed to
+/// run it. it is the top level object on which to invoke operations.
+class Runner {
+private:
+  constexpr static const char *Message = "you can only registers containers "
+                                         "before adding a step, otherwise the "
+                                         "already present steps will not be "
+                                         "aware of the newly registered "
+                                         "containers.";
+
+  Context *TheContext;
+  ContainerFactorySet ContainerFactoriesRegistry;
+  bool IsContainerFactoriesRegistryFinalized = false;
+  Structure State;
```

Comment:
A `Runner` at this point seems like a verbose wrapper around `Structure`, can we squash `Structure` within `Runner`?